# Filter Data Using Pandas

In this activity, we will demonstrate how you can read in the data, select specific rows and columns, build conditions to filter for a subset of the dataframe, all using Pandas. 

## Reading a CSV file

We'll use the function `read_csv()` to load the data into our notebook

- The `read_csv()` function can read data from a locally saved file or from a URL
- We'll store the data as a variable `world`

In [1]:
import pandas as pd

world = pd.read_csv('gapminder.csv')
world.head()

country  year  population region     sub_region income_group  \
0  Afghanistan  1950     7750000   Asia  Southern Asia          Low   
1  Afghanistan  1955     8270000   Asia  Southern Asia          Low   
2  Afghanistan  1960     9000000   Asia  Southern Asia          Low   
3  Afghanistan  1965     9940000   Asia  Southern Asia          Low   
4  Afghanistan  1970    11100000   Asia  Southern Asia          Low   

   life_expectancy  gdp_per_capita  children_per_woman  child_mortality  \
0             32.0            1040                7.57            425.0   
1             35.1            1130                7.52            394.0   
2             38.6            1210                7.45            364.0   
3             42.2            1190                7.45            334.0   
4             45.8            1180                7.45            306.0   

   pop_density  years_in_school_men  years_in_school_women  
0         11.9                  NaN                    NaN  
1         12.7                  NaN                    NaN  
2         13.8                  NaN                    NaN  
3         15.2                  NaN                    NaN  
4         17.0                 1.36                   0.21

- The output is truncated but the data is all there in our `world` variable
- Each row of the table is an **observation**, containing the data for a single country in a single year
- You may notice some weird `NaN` values&mdash;these represent missing data (`NaN` = "not a number")

What type of object is `world`?

In [2]:
type(world)

pandas.core.frame.DataFrame

We can access the `columns` attribute to print out all the columns in the dataframe. This is especially useful whenever you have a very wide dataframe.

In [3]:
world.columns

Index(['country', 'year', 'population', 'region', 'sub_region', 'income_group',
       'life_expectancy', 'gdp_per_capita', 'children_per_woman',
       'child_mortality', 'pop_density', 'years_in_school_men',
       'years_in_school_women'],
      dtype='object')

## Selecting Columns

Similar to a Python dictionary, we can index a specific column of a DataFrame using the column name inside square brackets. 

To select a single column, type the name of the column inside square brackets.

In [5]:
world['year']

0       1950
1       1955
2       1960
3       1965
4       1970
        ... 
2487    1995
2488    2000
2489    2005
2490    2010
2491    2015
Name: year, Length: 2492, dtype: int64

When you select one column from a dataframe, the output is a Series.

In [4]:
type(world['year'])

pandas.core.series.Series

To Select multiple columns, put a column names inside a Python list, and put that list inside a pair of square brackets:

In [5]:
world[['country', 'year']]

country  year
0     Afghanistan  1950
1     Afghanistan  1955
2     Afghanistan  1960
3     Afghanistan  1965
4     Afghanistan  1970
...           ...   ...
2487     Zimbabwe  1995
2488     Zimbabwe  2000
2489     Zimbabwe  2005
2490     Zimbabwe  2010
2491     Zimbabwe  2015

[2492 rows x 2 columns]

Note the double square brackets!
- These are required because we need both:
  - A pair of square brackets to extract the subset, AND
  - A pair of square brackets to define the list of columns to select.

When you select more than one column, the output is a DataFrame:

In [6]:
type(world[['year', 'country']])

pandas.core.frame.DataFrame

It's fine to type a list of column names in order to get a subset of the dataframe. However, this is time-consuming. If you'll be frequently using a particular subset, it's often helpful to assign the subset dataframe to a separate variable.

In [7]:
populations = world[['country', 'year', 'population']]
populations.head()

country  year  population
0  Afghanistan  1950     7750000
1  Afghanistan  1955     8270000
2  Afghanistan  1960     9000000
3  Afghanistan  1965     9940000
4  Afghanistan  1970    11100000

When selecting a larger number of columns, it may also be helpful to assign the list of column names to a separate variable.

In [8]:
subset_columns = ['country', 'region', 'year', 'population', 'life_expectancy']
world_subset = world[subset_columns]
world_subset.head()

country region  year  population  life_expectancy
0  Afghanistan   Asia  1950     7750000             32.0
1  Afghanistan   Asia  1955     8270000             35.1
2  Afghanistan   Asia  1960     9000000             38.6
3  Afghanistan   Asia  1965     9940000             42.2
4  Afghanistan   Asia  1970    11100000             45.8

### Other methods to select rows and data
#### Slicing and Indexing

**Slicing:**
- Select specific ranges of our data in both the row and column directions using either label or integer-based indexing.

In [9]:
world_subset[0:3]

country region  year  population  life_expectancy
0  Afghanistan   Asia  1950     7750000             32.0
1  Afghanistan   Asia  1955     8270000             35.1
2  Afghanistan   Asia  1960     9000000             38.6

**Indexing:**

> - `iloc` is integer based indexing.

In [10]:
world_subset.iloc[0:3, 1:4]

region  year  population
0   Asia  1950     7750000
1   Asia  1955     8270000
2   Asia  1960     9000000

* `loc` is label based indexing. Integers may be used but they are interpreted as a label.
* The following code will select rows of index value 0 and 10, and select all the columns.

In [11]:
# Select all columns for rows of index values 0 and 10
world_subset.loc[[0, 10], :]

country region  year  population  life_expectancy
0   Afghanistan   Asia  1950     7750000             32.0
10  Afghanistan   Asia  2000    20100000             51.6

* The following code will select row of index 0, and only the columns of region, life_expectancy, and year.

In [12]:
world_subset.loc[0, ['region', 'life_expectancy', 'year']]

region             Asia
life_expectancy    32.0
year               1950
Name: 0, dtype: object

* The following code will select rows of index values 1, 2, and 3.

In [13]:
world_subset.iloc[1:4]

country region  year  population  life_expectancy
1  Afghanistan   Asia  1955     8270000             35.1
2  Afghanistan   Asia  1960     9000000             38.6
3  Afghanistan   Asia  1965     9940000             42.2

### Unique Values in a Column

If we just want a list of unique values, we can use the `unique()` method:

In [14]:
world['year'].unique()

array([1950, 1955, 1960, 1965, 1970, 1975, 1980, 1985, 1990, 1995, 2000,
       2005, 2010, 2015])

If we just want to know how many unique values there are in a column, we can use the `nunique()` method:

In [ ]:
world['sub_region'].nunique()

### Sorting Values

A helpful way to summarize categorical data (such as names of countries and regions) is use the `value_counts()` method to count the unique values in that column:


In [6]:
world['sub_region'].value_counts()

Sub-Saharan Africa                 644
Latin America and the Caribbean    406
Western Asia                       252
Southern Europe                    168
Eastern Europe                     140
Northern Europe                    140
South-eastern Asia                 140
Southern Asia                      126
Western Europe                      98
Northern Africa                     84
Eastern Asia                        70
Central Asia                        70
Melanesia                           56
Australia and New Zealand           28
Northern America                    28
Polynesia                           28
Micronesia                          14
Name: sub_region, dtype: int64

The output above tells us, for example, that 644 of the observations in our data are for Sub-Saharan Africa (recall that each row is an observation corresponding to a single country in a single year).

- By default, `value_counts()` sorts the output from highest count to lowest
- To sort by the sub-region name, we can chain the `sort_index()` method

In [16]:
world.sort_values(by='year', ascending=True)

country  year  population    region  \
0       Afghanistan  1950     7750000      Asia   
938           Haiti  1950     3220000  Americas   
924          Guyana  1950      407000  Americas   
910   Guinea-Bissau  1950      535000    Africa   
896          Guinea  1950     3090000    Africa   
...             ...   ...         ...       ...   
1511        Morocco  2015    34800000    Africa   
1497     Montenegro  2015      628000    Europe   
1483       Mongolia  2015     2980000      Asia   
1455         Mexico  2015   126000000  Americas   
2491       Zimbabwe  2015    15800000    Africa   

                           sub_region  income_group  life_expectancy  \
0                       Southern Asia           Low             32.0   
938   Latin America and the Caribbean           Low             35.3   
924   Latin America and the Caribbean  Upper middle             58.0   
910                Sub-Saharan Africa           Low             36.2   
896                Sub-Saharan Africa           Low             38.5   
...                               ...           ...              ...   
1511                  Northern Africa  Lower middle             74.8   
1497                  Southern Europe  Upper middle             76.9   
1483                     Eastern Asia  Lower middle             68.2   
1455  Latin America and the Caribbean  Upper middle             76.2   
2491               Sub-Saharan Africa           Low             58.3   

      gdp_per_capita  children_per_woman  child_mortality  pop_density  \
0               1040                7.57            425.0        11.90   
938             2290                6.31            352.0       117.00   
924             4060                5.48            155.0         2.07   
910              837                7.09            339.0        19.00   
896              614                5.98            364.0        12.60   
...              ...                 ...              ...          ...   
1511            7290                2.53             28.0        78.00   
1497           15300                1.68              4.1        46.70   
1483           11400                2.79             18.8         1.92   
1455           16700                2.22             15.0        64.80   
2491            1890                3.84             59.9        40.80   

      years_in_school_men  years_in_school_women  
0                     NaN                    NaN  
938                   NaN                    NaN  
924                   NaN                    NaN  
910                   NaN                    NaN  
896                   NaN                    NaN  
...                   ...                    ...  
1511                 6.83                   4.84  
1497                12.90                  13.10  
1483                10.80                  12.10  
1455                10.80                  10.80  
2491                11.10                  10.20  

[2492 rows x 13 columns]

## Selecting Rows

Similar to applying a filter in Excel, we can extract rows from a DataFrame or Series based on a criteria.

For example, suppose we want to select the rows where the life expectancy is greater than 82 years

First we use the comparison operator `>` on the `life_expectancy` column:

In [18]:
world['life_expectancy'] > 82

0       False
1       False
2       False
3       False
4       False
        ...  
2487    False
2488    False
2489    False
2490    False
2491    False
Name: life_expectancy, Length: 2492, dtype: bool

The result is a Series with a Boolean value for each row in the dataframe, indicating whether it is `True` or `False` -- whether a row has a value above 82 in the column `life_expectancy`.

We can find out how many rows match this condition, by using the `sum()` method
- In Python, the Boolean value of `True` has a value of 1, and `False` has a value of 0.

In [19]:
above_82 = world['life_expectancy'] > 82
above_82.sum()

13

We can use a Boolean Series (this is a Series ofo Booleans) as a **filter** to extract the rows of `world` which have life expectancy above 82
- Previously we used square brackets and a column name or list of column names to extract *column(s)* from a DataFrame (e.g. `df['X']`)
- Now we use square brackets and a Boolean Series to extract *rows* from a DataFrame

In [20]:
world[world['life_expectancy'] >= 82].head(5)

country  year  population   region                 sub_region  \
110  Australia  2010    22100000  Oceania  Australia and New Zealand   
111  Australia  2015    23800000  Oceania  Australia and New Zealand   
783     France  2015    64500000   Europe             Western Europe   
992    Iceland  2010      320000   Europe            Northern Europe   
993    Iceland  2015      330000   Europe            Northern Europe   

    income_group  life_expectancy  gdp_per_capita  children_per_woman  \
110         High             82.0           41400                1.93   
111         High             82.6           43800                1.86   
783         High             82.2           37800                1.98   
992         High             82.0           38800                2.07   
993         High             82.2           42700                1.95   

     child_mortality  pop_density  years_in_school_men  years_in_school_women  
110              4.8         2.88                 13.5                   13.8  
111              3.8         3.10                 14.2                   14.5  
783              3.9       118.00                 13.2                   14.0  
992              2.6         3.20                 13.3                   14.2  
993              2.2         3.29                 13.9                   14.9

- The output is a DataFrame containing only the 13 rows of `world` matching our criteria
- We can see which countries and years with life expectancy above 82 are:
  - Australia, France, Iceland, Italy, Japan, Norway, Singapore, Spain, Sweden and Switzerland in 2015
  - Japan, Singapore, and Switzerland in 2010

We can use any of the comparison operators (`>`, `>=`, `<`, `<=`, `==`, `!=`) on a DataFrame column to create Boolean Series for filtering our data

Select data for East Asian countries:

In [21]:
east_asia = world[world['sub_region'] == 'Eastern Asia']
east_asia.head(3)

country  year  population region    sub_region  income_group  \
448   China  1950   554000000   Asia  Eastern Asia  Upper middle   
449   China  1955   611000000   Asia  Eastern Asia  Upper middle   
450   China  1960   658000000   Asia  Eastern Asia  Upper middle   

     life_expectancy  gdp_per_capita  children_per_woman  child_mortality  \
448             40.7             536                5.29            317.0   
449             47.0             708                5.98            291.0   
450             30.9             891                3.99            309.0   

     pop_density  years_in_school_men  years_in_school_women  
448         59.1                  NaN                    NaN  
449         65.1                  NaN                    NaN  
450         70.1                  NaN                    NaN

We can filter the East Asian data further to select the year 2015:

In [22]:
east_asia_2015 = east_asia[east_asia['year'] == 2015]
east_asia_2015.head(3)

country  year  population region    sub_region  income_group  \
461      China  2015  1400000000   Asia  Eastern Asia  Upper middle   
1119     Japan  2015   128000000   Asia  Eastern Asia          High   
1483  Mongolia  2015     2980000   Asia  Eastern Asia  Lower middle   

      life_expectancy  gdp_per_capita  children_per_woman  child_mortality  \
461              76.3           13600                1.62             10.7   
1119             83.8           37800                1.44              3.0   
1483             68.2           11400                2.79             18.8   

      pop_density  years_in_school_men  years_in_school_women  
461        149.00                 11.0                   9.82  
1119       351.00                 15.1                  15.50  
1483         1.92                 10.8                  12.10

## Selecting Rows and Columns
### How to Access Data in the table? 

To select rows and columns at the same time, we use the syntax `.loc[<rows>, <columns>]`:

In [23]:
canada_pop = world.loc[world['country'] == 'Canada', ['country', 'year', 'population']]
canada_pop.head()

country  year  population
392  Canada  1950    13700000
393  Canada  1955    15700000
394  Canada  1960    17900000
395  Canada  1965    19700000
396  Canada  1970    21500000

In [24]:
random_pop = world.loc[392:396, 'country':'population']
random_pop

country  year  population
392  Canada  1950    13700000
393  Canada  1955    15700000
394  Canada  1960    17900000
395  Canada  1965    19700000
396  Canada  1970    21500000

- To improve readability, longer statements can be split into multiple rows

In [25]:
recent_data = world.loc[world['year'] >= 2010,
                        ['country', 'year', 'sub_region', 'population', 'life_expectancy']
                       ]
recent_data.head()

country  year       sub_region  population  life_expectancy
12  Afghanistan  2010    Southern Asia    28800000             56.2
13  Afghanistan  2015    Southern Asia    33700000             57.9
26      Albania  2010  Southern Europe     2940000             76.3
27      Albania  2015  Southern Europe     2920000             77.6
40      Algeria  2010  Northern Africa    36100000             76.5

## More Options for Data Selection

- A single expression can be used to filter rows based on multiple criteria, either matching *all* the criteria (`&`) or *any* of the criteria (`|`)
- These special operators are used instead of `and` and `or` to make sure that the comparison occurs for each row in the data frame
- Parentheses are added to indicate the priority of the comparisons

Select rows where the sub-region is Northern Europe and the year is 2015:

In [26]:
world.loc[(world['sub_region'] == 'Northern Europe') & (world['year'] == 2015),
          ['sub_region', 'country', 'year', 'gdp_per_capita']
         ].head(3)

sub_region  country  year  gdp_per_capita
615  Northern Europe  Denmark  2015           45500
727  Northern Europe  Estonia  2015           27300
769  Northern Europe  Finland  2015           39000

Other useful ways of subsetting data include methods such as `isin()`, `between()`, `isna()`, `notna()`

In [27]:
world.loc[world['country'].isin(['Canada', 'Japan', 'France']) & (world['year'] == 2015),
          ['country', 'year', 'life_expectancy']
         ]

country  year  life_expectancy
405   Canada  2015             81.7
783   France  2015             82.2
1119   Japan  2015             83.8

Want to learn more? Check out [this tutorial](https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html#min-tut-03-subset) and the [pandas documentation](https://pandas.pydata.org/docs/user_guide/indexing.html).